api/src/migrations/geoprocessing/1611221157285-initialGeoDBSetup.ts

```diff
@@ -0,0 +1,185 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class initialGeoDBSetup1611221157285 implements MigrationInterface {
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+      await queryRunner.query(`
+      CREATE EXTENSION IF NOT EXISTS "uuid-ossp";
+      CREATE EXTENSION IF NOT EXISTS tablefunc;
+      CREATE EXTENSION IF NOT EXISTS plpgsql;
+      CREATE EXTENSION IF NOT EXISTS postgis;
+      CREATE EXTENSION IF NOT EXISTS postgis_raster; -- OPTIONAL
+      CREATE EXTENSION IF NOT EXISTS postgis_topology; -- OPTIONAL
+
+      CREATE TYPE "source_type" AS ENUM (
+        'user_imported',
+        'gbif',
+        'uicn',
+        'ecoregions',
+        'intersection'
+      );
+
+      CREATE TYPE "shape_type" AS ENUM (
+        'square',
+        'hexagon',
+        'irregular'
+      );
+
+      CREATE TYPE "status" AS ENUM (
+        'created',
+        'running',
+        'done',
+        'failure'
+      );
+
+      CREATE TYPE "adm_level" AS ENUM (
+        0,
+        1,
+        2
+      );
+
+-- Administrative regions table.
+      CREATE TABLE "admin_regions" (
+        "id" uuid PRIMARY KEY DEFAULT uuid_generate_v4(),
+        "ogc_fid" int NOT NULL,
+        "the_geom" geometry CONSTRAINT admin_regions_geometry_valid_check CHECK (ST_IsValid(the_geom)),
+        "name_0" varchar,
+        "name_1" varchar,
+        "name_2" varchar,
+        "iso3" varchar(3),
+        "gid_0" varchar,
+        "gid_1" varchar,
+        "gid_2" varchar,
+        "level" adm_level NOT NULL,
+        "created_at" timestamp NOT NULL default now(),
+        "created_by" uuid NOT NULL,
+        "modify_at" timestamp NOT NULL default now()
+      ) PARTITION BY LIST(level);
+
+      CREATE TABLE admin_regions_0 PARTITION OF admin_regions FOR VALUES IN (0);
+      CREATE TABLE admin_regions_1 PARTITION OF admin_regions FOR VALUES IN (1);
+      CREATE TABLE admin_regions_2 PARTITION OF admin_regions FOR VALUES IN (2);
```

Comment:
💓

Reply:
this partition won't work as is now because of limitations to PostgreSQL partitions: https://www.postgresql.org/docs/13/ddl-partitioning.html#DDL-PARTITIONING-DECLARATIVE-LIMITATIONS

Specifically:
> Unique constraints (and hence primary keys) on partitioned tables must include all the partition key columns. This limitation exists because PostgreSQL can only enforce uniqueness in each partition individually.

so one way would be to use `PRIMARY KEY(id, level)` - in theory this could mean that a clash of ids across different partitions would prevent partitions from being joined later on, but since `id` is UUIDv4 this may happen after the end of the universe...

or we could switch to partitioning by hash, but I don't think this makes sense here